server/api/services/service.ts

```diff
@@ -34,72 +34,93 @@ export class ServiceProviderService {
+      params.append("grant_type", "authorization_code");
+      params.append("client_id", process.env.ASANA_CLIENT_ID);
+      params.append("client_secret", process.env.ASANA_CLIENT_SECRET);
+      params.append("code", unescape(code));
+      params.append("redirect_uri", redirect_uri);
 
-      
-      */
+      fetch(process.env.ASANA_TOKEN_EXCHANGE_URL, {
+        method: "post",
+        body: params,
+      })
+        .then((res) => res.json())
+        .then((response) => {
+          if (response.error)
+            reject({ code: 500, message: response.error_description });
+          else resolve(response.access_token);
+        });
     });
   }
 
-  static linkGithub(code: string) {
+  static linkDiscord(code: string, redirect_uri: string) {
     return new Promise((resolve, reject) => {
       const params = new URLSearchParams();
-      params.append("client_id", process.env.GITHUB_CLIENT_ID);
-      params.append("client_secret", process.env.GITHUB_CLIENT_SECRET);
+      params.append("client_id", process.env.DISCORD_CLIENT_ID);
+      params.append("client_secret", process.env.DISCORD_CLIENT_SECRET);
+      params.append("grant_type", "authorization_code");
+      params.append("redirect_uri", redirect_uri);
       params.append("code", code);
+      params.append("scope", process.env.DISCORD_SCOPE);
 
-      fetch(process.env.GITHUB_TOKEN_EXCHANGE_URL, {
+      fetch(process.env.DISCORD_TOKEN_EXCHANGE_URL, {
         method: "post",
         body: params,
-        headers: { Accept: "application/json" },
+        headers: { "Content-Type": "application/x-www-form-urlencoded" },
       })
         .then((res) => res.json())
         .then((response) => {
+          if (response.redirect_uri)
+            reject({ code: 400, message: response.redirect_uri[0] });
           if (response.error)
             reject({ code: 500, message: response.error_description });
           else resolve(response.access_token);
         });
     });
   }
 
-  static linkHeroku(code: string) {
+  static linkGithub(code: string) {
     return new Promise((resolve, reject) => {
       const params = new URLSearchParams();
-      params.append("grant_type", "authorization_code");
+      params.append("client_id", process.env.GITHUB_CLIENT_ID);
+      params.append("client_secret", process.env.GITHUB_CLIENT_SECRET);
       params.append("code", code);
-      params.append("client_secret", process.env.HEROKU_CLIENT_SECRET);
 
-      fetch(process.env.HEROKU_TOKEN_EXCHANGE_URL, {
+      fetch(process.env.GITHUB_TOKEN_EXCHANGE_URL, {
         method: "post",
         body: params,
+        headers: { Accept: "application/json" },
       })
         .then((res) => res.json())
         .then((response) => {
-          if (response.id == "unauthorized")
-            reject({ code: 500, message: response.message });
+          if (response.error)
+            reject({ code: 500, message: response.error_description });
```

Comment:
I think we should send a proper message here instead of just url

Reply:
There is a proper message here. `response.error_description` contains the message automatically, from the service.